### DISCUSSION TITLE
Why does LazyServiceProvider use a cache?

### DISCUSSION THREAD

**Discussion by damien-r** [abpframework/abp#14157] [General]
I have a project where I need different instances of `DomainService` implementations to manage different configurations.

At first, I used an injected `IServiceProvider` to generate my instance when needed. It worked.

After a refactor, I decided to use the provided property `Volo.Abp.Domain.Services.DomainService.ServiceProvider` to avoid having to inject a `IServiceProvider`. This provided property is however `Obsolete` and tells we need to use `LazyServiceProvider`.

But `LazyServiceProvider` uses a cache (`AbpLazyServiceProvider : CachedServiceProviderBase`). Why does it use a cache, I can't get different instance of the same service to manage different data. If it is by design, why is there no alternative to avoid using a cache?

**Comment by maliming** [comment]
> I have a project where I need different instances of DomainService implementations to manage different configurations.

In this case, you can resolve `IServiceProvider ` from `LazyServiceProvider `

**Reply by damien-r** [reply]
> > I have a project where I need different instances of DomainService implementations to manage different configurations.
> 
> In this case, you can resolve `IServiceProvider ` from `LazyServiceProvider `

Thanks for your reply. I can but it looks like a work around.

I find counter-intuitive to declare object as Transient: `DomainService` is Transient (https://docs.abp.io/en/abp/latest/Dependency-Injection#inherently-registered-types) and to get the same instance because of a cache. IMHO, this cache usage is incorrect for Transient.

**Reply by damien-r** [reply]
> > I have a project where I need different instances of DomainService implementations to manage different configurations.
> 
> In this case, you can resolve `IServiceProvider ` from `LazyServiceProvider `

And why is the provided instance obsolete if you advice to inject it anyway, but indirectly through `LazyServiceProvider `?
